specs/merge/beacon-chain.md

```diff
@@ -0,0 +1,190 @@
+
+```python
+class Transaction(Container):
+    nonce: uint64
+    gas_price: uint256
+    gas_limit: uint64
+    recipient: Bytes20
+    value: uint256
+    input: List[Bytes1, MAX_BYTES_PER_TRANSACTION_PAYLOAD]
+    v: uint256
+    r: uint256
+    s: uint256
+```
+
+#### `ApplicationPayload`
+
+The application payload included in a `BeaconBlock`.
+
+```python
+class ApplicationPayload(Container):
+    block_hash: Bytes32  # Hash of application block
+    coinbase: Bytes20
+    state_root: Bytes32
+    gas_limit: uint64
+    gas_used: uint64
+    receipt_root: Bytes32
+    logs_bloom: Vector[Bytes1, BYTES_PER_LOGS_BLOOM]
+    difficulty: uint64  # Temporary field, will be removed later on
+    transactions: List[Transaction, MAX_APPLICATION_TRANSACTIONS]
+```
+
+## Helper functions
+
+### Misc
+
+#### `is_transition_completed`
+
+```python
+def is_transition_completed(state: BeaconState) -> boolean:
+    state.application_block_hash != Bytes32()
+```
+
+#### `is_transition_block`
+
+```python
+def is_transition_block(state: BeaconState, block_body: BeaconBlockBody) -> boolean:
+    return state.application_block_hash == Bytes32() and block_body.application_payload.block_hash != Bytes32()
+```
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    process_block_header(state, block)
+    process_randao(state, block.body)
+    process_eth1_data(state, block.body)
+    process_operations(state, block.body)
+    process_application_payload(state, block.body)  # [New in Merge]
+```
+
+#### Application payload processing
+
+##### `get_application_state`
+
+*Note*: `ApplicationState` class is an abstract class representing ethereum application state.
+
+Let `get_application_state(application_state_root: Bytes32) -> ApplicationState`  be the function that given the root hash returns a copy of ethereum application state. 
+The body of the function is implementation dependent.
+
+##### `application_state_transition`
+
+Let `application_state_transition(application_state: ApplicationState, application_payload: ApplicationPayload) -> None` be the transition function of ethereum application state. 
+The body of the function is implementation dependent.
+
+*Note*: `application_state_transition` must throw `AssertionError` if either the transition itself or one of the post-transition verifications has failed.
+
+##### `process_application_payload`
+
+```python
+def process_application_payload(state: BeaconState, body: BeaconBlockBody) -> None:
+    """
+    Note: This function is designed to be able to be run in parallel with 
+    the other `process_block` sub-functions
+    """
+
+    if is_transition_completed(state):
+        application_state = get_application_state(state.application_state_root)
+        application_state_transition(application_state, body.application_payload)
+
+        state.application_state_root = body.application_payload.state_root
```

Comment:
Is there a need for the Eth2 node to access `application_state`? 
A cleaner approach would be to outsource all the work to `application_state_transition()`, so that it operates using the `application_state_root`. (see suggestion for the `application_state_transition` function)
Also, need to check that whatever is returned by `application_state_transition()` actually matches with `body.application_payload.state_root`.

```suggestion
        state.application_state_root = application_state_transition(state.application_state_root, body.application_payload)
        assert state.application_state_root == body.application_payload.state_root
```

Reply:
I definitely see the value in simplifying this part. But then it would not read like the beacon state and the application state are tightly coupled.

For the root check we can do the following:
```python
application_state = get_application_state(state.application_state_root)
application_state_transition(application_state, body.application_payload)

assert body.application_payload.state_root == get_application_state_root(application_state)

state.application_state_root = body.application_payload.state_root
state.application_block_hash = body.application_payload.block_hash
```

But that would require yet another abstract function. So, it might really be better to have the following (according to your suggestion):
```python
application_state_root = application_state_transition(state.application_state_root, body.application_payload)

assert application_state_root == body.application_payload.state_root

state.application_state_root = body.application_payload.state_root
state.application_block_hash = body.application_payload.block_hash
```

The latter form also fits stateless verification.

Reply:
nitpicking: `application_state_root = application_state_transition(...)` requires `application_state_transition` to return `Bytes32`, but it doesn't align with the naming pattern as beacon state transition function `state_transition(...) -> None`. So having a `get_application_state_root(application_state)` makes sense to me.

Reply:
Making this call conformant with `state_transition(...)` was the original intention put behind this extra function call and using application state object rather than the state root.

Reply:
We may also define `ApplicationState` in a following way
```python
class ApplicationState(Container):
    root: Bytes32
```

And then add an explicit check of the state root after application state transition
```python
application_state = get_application_state(state.application_state_root)
application_state_transition(application_state, body.application_payload)

assert application_state.root == body.application_payload.state_root

state.application_state_root = body.application_payload.state_root
state.application_block_hash = body.application_payload.block_hash
```